### PR TITLE
Add Traits

### PR DESCRIPTION
**Pull request by kazk**
A quick hack to add traits. It compiles :)

### CODE REVIEW

File: justfile [hunk 1]
```diff
@@ -61,21 +61,8 @@ swagger-transform:
 # Download and generate all swagger dependent files
 swagger: swagger-dl swagger-patch swagger-transform
 
-# Build a FileDescriptorSet for custom code generation
```
Comment (kazk) [inline justfile:64]: Wasn't necessary because `prost_build` can generate it.

File: k8s-pb-codegen/src/main.rs [hunk 2]
```diff
@@ -95,5 +107,172 @@ fn main() -> Result<()> {
...
+                fn metadata_mut(&mut self) -> Option<&mut<Self as crate::HasMetadata>::Metadata> {
+                    self.metadata.as_mut()
+                }
+            }
+        }
+    } else {
+        tokens
+    };
+    let tokens = if let Some(spec) = &resource.spec {
+        // HACK Prost changes message name `APIService` to `ApiService`, and `transformed.json` uses the original name.
```
Comment (kazk) [inline k8s-pb-codegen/src/main.rs:195]: Prost uses `heck` to transform message name, so some of them doesn't match the name extracted from `swagger.json`.

File: k8s-pb-codegen/src/main.rs [hunk 3]
```diff
@@ -50,18 +54,26 @@ fn main() -> Result<()> {
...
+                let message_name = match msg.name {
+                    Some(ref name) => name,
+                    None => continue,
+                };
+                let path = format!("{}.{}", package_name, message_name);
+                let resource = match resources.get(&path) {
+                    Some(resource) => resource,
+                    None => continue,
+                };
+                append_trait_impl(&mut pkg_rs, message_name, resource);
```
Comment (clux) [inline k8s-pb-codegen/src/main.rs:73]: the `continue` here goes to the innermost loop, so this is equivalent to:
```suggestion
                if let Some(resource) = resources.get(&path) {
                    append_trait_impl(&mut pkg_rs, message_name, resource);
                }
```

File: k8s-pb-codegen/src/main.rs [hunk 4]
```diff
@@ -95,5 +107,172 @@ fn main() -> Result<()> {
...
+    let mut lib_rs = OpenOptions::new()
+        .append(true)
+        .open(target_dir.join("lib.rs"))
+        .unwrap();
+    append_trait_def(&mut lib_rs);
+
     Ok(())
 }
+
+fn append_trait_def(lib_rs: &mut File) {
```
Comment (clux) [inline k8s-pb-codegen/src/main.rs:119]: hah, that is a nice way to get started. We can think about where to move this stuff later in https://github.com/kube-rs/k8s-pb/issues/4

File: k8s-pb-codegen/src/main.rs (hunk 2 above)
Comment (clux) [reply k8s-pb-codegen/src/main.rs:195]: That is nicer, I think. Although it creates the problem of us having to disambiguate `APIResource` with `ApiResource`, but that might be something we ought to resolve in another way.

File: k8s-pb-codegen/src/main.rs [hunk 5]
```diff
@@ -95,5 +107,172 @@ fn main() -> Result<()> {
...
+        tokens
+    };
+
+    let tokens = rustfmt(tokens).unwrap();
+    writeln!(pkg_rs).unwrap();
+    writeln!(pkg_rs, "{}", &tokens).unwrap();
+}
+
+// Run `rustfmt` on an in-memory string
+fn rustfmt(text: impl std::fmt::Display) -> Result<String, Box<dyn std::error::Error>> {
```
Comment (clux) [inline k8s-pb-codegen/src/main.rs:268]: Ah this is nice.

**Review by clux — APPROVED** [review]
this is great. it allows us to start iterating on this properly.
one minor comment, but otherwise am just super happy to see this :+1:

File: k8s-pb-codegen/src/main.rs (hunk 2 above)
Comment (kazk) [reply k8s-pb-codegen/src/main.rs:193]: Yeah, makes them fit Rust's naming convention better. But I'm not sure if we should start using different casing from the official ones. We can worry about it later.

File: k8s-pb-codegen/src/main.rs [hunk 6]
```diff
@@ -95,5 +105,172 @@ fn main() -> Result<()> {
...
+
+            impl crate::HasConditions for #type_name {
+                type Condition = crate::#(#condition)::*;
+
+                fn conditions(&self) -> Option<&[<Self as crate::HasConditions>::Condition]> {
+                    self.status.as_ref().map(|s| s.conditions.as_slice())
+                }
+                fn conditions_mut(&mut self) -> Option<&mut Vec<<Self as crate::HasConditions>::Condition>> {
+                    self.status.as_mut().and_then(|s| Some(s.conditions.as_mut()))
+                }
```
Comment (clux) [inline k8s-pb-codegen/src/main.rs:253]: i love how clean even this is.

File: k8s-pb-codegen/src/main.rs (hunk 2 above)
Comment (clux) [reply k8s-pb-codegen/src/main.rs:193]: yeah, i'll open an issue to track it

File: k8s-pb-codegen/src/main.rs [hunk 7]
```diff
@@ -95,5 +105,172 @@ fn main() -> Result<()> {
...
     Ok(())
 }
+
+fn append_trait_def(lib_rs: &mut File) {
+    let tokens = quote! {
+        pub trait Resource {
+            const API_VERSION: &'static str;
+            const GROUP: &'static str;
+            const KIND: &'static str;
+            const VERSION: &'static str;
```
Comment (kazk) [inline k8s-pb-codegen/src/main.rs:123]: Ugh, GKV. This bothers me, so let me fix and regenerate. I'll merge this after that.

**Pull request merged**